scipy/signal/_signaltools.py

```diff
@@ -4625,3 +4625,128 @@ def decimate(x, q, n=None, ftype='iir', axis=-1, zero_phase=True):
+
+    Notes
+    -----
+    .. versionadded:: 1.12.0
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy.signal import envelope
+    >>> import matplotlib.pyplot as plt
+
+    Create a signal and calculate its analytical envelopes.
+
+    >>> t = np.arange(0, 2, 1/2000)
+    >>> signal = (1 + 0.5 * np.cos(2 * np.pi * 1 * t)) * np.cos(2 * np.pi * 10 * t) + 5
+    >>> upper, lower = envelope(signal)
+    >>> plt.plot(t, signal)
+    >>> plt.plot(t, upper)
+    >>> plt.plot(t, lower)
+    >>> plt.show()
+
+    Generate an amplitude modulated Gaussian white noise and calculate its
+    root-mean-square envelopes.
+
+    >>> t = np.arange(0, 2, 1/2000)
+    >>> gaussian_white_noise = np.random.normal(size=t.shape[0])
+    >>> signal = (1 + 0.5 * np.cos(2 * np.pi * 1 * t)) * gaussian_white_noise + 5
+    >>> upper, lower = envelope(signal, N=150, method='rms')
+    >>> plt.plot(t, signal)
+    >>> plt.plot(t, upper)
+    >>> plt.plot(t, lower)
+    >>> plt.show()
+
+    Raises
+    ------
+    ValueError
+        If the input array is not 1D.
+        If `N` is not a positive integer.
+        If `N` is not specified when using the `'rms'` or `'peak'` method.
+        If the specified method is not valid.
+    """
+    x = np.squeeze(np.asarray(x))
+    if x.ndim != 1:
+        raise ValueError("Input array must be a 1D array")
+    if N is not None and (not isinstance(N, int) or N <= 0):
+        raise ValueError('If N is not None, it must be a positive integer.')
+    if method == 'analytic':
+        x_mean = np.mean(x)
+        x_zero_mean = x - x_mean
+        zero_mean_envelope = np.abs(hilbert(x_zero_mean, N=N))
+        return x_mean + zero_mean_envelope, x_mean - zero_mean_envelope
+    elif method == 'rms':
+        if N is None:
+            raise ValueError('N cannot be None when using the rms method; '
+                             'it must specify the sliding window length.')
+        x_mean = np.mean(x)
+        x_zero_mean = x - x_mean
+        zero_mean_envelope = np.sqrt(convolve(x_zero_mean**2, np.ones(N)/N,
+                                              mode='same', method='direct'))
+        return x_mean + zero_mean_envelope, x_mean - zero_mean_envelope
+    elif method == 'peak':
+        # Doing this import at the top of the file throws a circular import
+        # error. One way to avoid this error is doing a local import as
+        # implemented here. Another way is to move this function to
+        # _peak_finding and importing hilbert,convolve and UnivariateSpline
+        # there.
+        from ._peak_finding import find_peaks
+
+        if N is None:
+            msg = ('N cannot be None when using the peak method; it must specify '
+                   'the minimal horizontal distance between neighbouring peaks')
+            raise ValueError(msg)
+        peaks_upper, _ = find_peaks(x, distance=N)
+        peaks_lower, _ = find_peaks(-x, distance=N)
```

Comment:
I don't have much experience with using peaks to detect an envelope but maybe the `prominence` could also be a useful constraint to select significant peaks for the envelope? One could add `**kwargs` or `peak_kw` as a parameter's which are passed to `find_peaks`.